provider/vsphere/internal/vsphereclient/createvm.go

```diff
@@ -402,24 +402,48 @@ func (c *Client) selectDatastore(
 	if err := c.client.Retrieve(ctx, refs, nil, &datastores); err != nil {
 		return nil, errors.Annotate(err, "retrieving datastore details")
 	}
+
+	if len(datastores) == 0 {
+		c.logger.Errorf("no datastores available")
+		return nil, errors.New("no datastores available")
+	}
+
 	if args.Constraints.RootDiskSource != nil {
 		dsName := *args.Constraints.RootDiskSource
 		c.logger.Debugf("desired datasource %q", dsName)
 		for _, ds := range datastores {
-			c.logger.Debugf("seen %q", ds.Name)
+			c.logger.Debugf("seen %s", ds.Name)
```

Comment:
There's not enough context in this debug message.
It should probably be Tracef as well. What really needs doing is that we record the names in a slice and include the names in the error message below.
could not find datastore %q from %q, dsName, availaableDatastores

Reply:
This is in a loop, so I don't know if we want to print `accessibleDatastores` every time. Also this opens us up to the concern that @babbageclunk raised about cases where there might be hundreds of datastores.